ran-core/src/main/java/io/ran/ClazzMethod.java

```diff
@@ -83,12 +92,61 @@ public boolean isSynchronized() {
 		return Modifier.isSynchronized(method.getModifiers());
 	}
 
+	private List<Type> getAllGenericTypes() {
+		ArrayList<Type> types = new ArrayList<>();
+		types.add(method.getGenericReturnType());
+		types.addAll(Arrays.asList(method.getGenericParameterTypes()));
+		types.addAll(Arrays.asList(method.getGenericExceptionTypes()));
+		return types.stream().flatMap(this::flattenGenericTypes).collect(Collectors.toList());
+	}
+
+	private Stream<Type> flattenGenericTypes(Type type) {
+		if (type instanceof ParameterizedType) {
+			return Arrays.stream(((ParameterizedType) type).getActualTypeArguments()).flatMap(this::flattenGenericTypes);
+		}
+		if (type instanceof GenericArrayType) {
+			return flattenGenericTypes(((GenericArrayType) type).getGenericComponentType());
+		}
+		return Stream.of(type);
+	}
+
+	/**
+	 * @return true when the method either returns or takes generic parameters defined at class-level. Method level parameter containing class level parameter does not count.
+	 */
+	public boolean hasGenericFromClass() {
+		for (Type type : getAllGenericTypes()) {
+			if (type instanceof TypeVariable) {
+				if (((TypeVariable<?>) type).getGenericDeclaration() == method.getDeclaringClass()) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * @return true if the method either returns or takes generic parameters defined at  method-level.
+	 */
+	public boolean hasGenericFromMethod() {
+		for (Type type : getAllGenericTypes()) {
+			if (type instanceof TypeVariable) {
+				if (((TypeVariable<?>) type).getGenericDeclaration() == method)
+					return true;
+			} else if (type instanceof WildcardType) {
+				return true;
+			}
```

Comment:
Should mention in the JavaDocs that wildcards are always treated as method-level generic

Reply:
what javadocs :laughing:

Reply:
done